polytope/polytope.py

```diff
@@ -2059,78 +2063,93 @@ def region_diff(poly, reg, abs_tol=ABS_TOL, intersect_tol=ABS_TOL,
             if save:
                 logger.debug('counter[level] is 0')
 
+            # Go through all remaining polytopes that we want to remove and check whether any of them removes
+            # anything from the current set of hyperplanes (constraints)
             for j in xrange(level, N):
+                # Construct a set of constraints with the current hyperplanes and the hyperplanes of polytope j
+                # This is the intersection of the current hyperplanes and polytope j
                 auxINDICES = np.hstack([
                     INDICES,
                     range(beg_mi[j], beg_mi[j] + mi[j])
                 ])
                 Adummy = A[auxINDICES, :]
                 bdummy = B[auxINDICES]
-                R, xopt = cheby_ball(Polytope(Adummy, bdummy))
+                # Will polytope j remove anything from the current set of hyperplanes?
+                R, _ = cheby_ball(Polytope(Adummy, bdummy))
                 if R > abs_tol:
+                    # Yes, constrain the set of constraints with the negation of one of polytope j's hyperplanes
                     level = j
                     counter[level] = 1
-                    INDICES = np.hstack([INDICES, beg_mi[level] + M])
+                    INDICES = np.hstack([INDICES, beg_mi[level] + M])  # Offset M negates hyperplanes
                     break
             if R < abs_tol:
-                level = level - 1
-                res = union(res, Polytope(A[INDICES, :], B[INDICES]), False)
-                nzcount = np.nonzero(counter)[0]
-                for jj in xrange(len(nzcount) - 1, -1, -1):
-                    if counter[level] <= mi[level]:
-                        INDICES[len(INDICES) -
-                                1] = INDICES[len(INDICES) - 1] - M
-                        INDICES = np.hstack([
-                            INDICES,
-                            beg_mi[level] + counter[level] + M
-                        ])
-                        break
-                    else:
-                        counter[level] = 0
-                        INDICES = INDICES[0:m + sum(counter)]
-                        if level == -1:
-                            logger.debug('returning res from 1st point')
-                            return res
+                # Since no polytope will remove anything, the current set of hyperplanes must be in the result
+                res = union(res, reduce(Polytope(A[INDICES, :], B[INDICES])), False)
+                # None of the remaining polytopes removes anything
+                # Indicate that we are done at this level
+                counter[level] = mi[level]
+                # Add the polytope to the indices to get an empty intersection
+                # This will force the algorithm to pop out of this level and move on
+                INDICES = np.hstack([
+                    INDICES,
+                    range(beg_mi[level], beg_mi[level] + mi[level])
+                ])
```

Comment:
I will try to address the concerns in several parts. Let's first look at the `return` statement.

For the program to jump to the `else` statement on the removed line 2091 the value of `counter[level]` must necessarily be strictly greater than 1 (and usually more). This is so because `mi` can only have positive elements. There is only one line on which `counter[level]` can be increased past 1, and that is on removed line 2104:
```
counter[level] = counter[level] + 1
```
Now, it is not sufficient for `counter[level]` to be strictly greater than 1 for the program to reach the `else` statement on line 2091. `counter[level]` must also be strictly greater than `mi[level]`. This can never happen at this point in the program. Right after the increment of `counter[level]` on line 2104 `counter[level]` is reset to 0 if `counter[level]´ is greater than `mi[level]`:
```
counter[level] = counter[level] + 1
if counter[level] <= mi[level]:
    ...
else:
    counter[level] = 0
```
Hence, the `else` statement on line 2091 cannot be reached and the return statement is dead code. I think this is why the indexing error on line 2093 has gone unnoticed.

Reply:
Let's also look at the `for` loop on removed line 2082. The iteration variable `jj` is never used, so we effectively have the following code on removed lines 2079-2096:
```
level = level - 1
res = union(res, Polytope(A[INDICES, :], B[INDICES]), False)
INDICES[-1] -= M
INDICES = np.hstack([
    INDICES,
    beg_mi[level] + counter[level] + M
])
```

Reply:
When running the test `test_checker_region_diff` in file `polytope_test.py` the values of some variables when the `while` loop on line 2071 is entered:
```
INDICES = [0, 1, 2, 3]
mi = [2, 2]
```
The content of `INDICES` represents the four sides of the minuend polytope. The values of `mi` tells us that the first and the second (ordered by decreasing Cheby ball radius) polytope in the subtrahend region has two hyperplanes each that are not found in the minuend polytope.

The halfplanes are indexed like this:
0-3: The four halfplanes of the minuend
4, 5: The two halfplanes of the first subtrahend
6, 7: The two halfplanes of the second subtrahend
8, 9: The complements of each of the halfplanes of the first subtrahend
10, 11: The complements of each of the halfplanes of the second subtrahend

First pass through while loop:
```
level == 0
counter == [0, 0]
```
The program is at first pass at level 0. The two hyperplanes (indices 4 and 5) of the largest subtrahend (happens to be the lower left box) is intersected with the minuend. The intersection is non-empty, so there is a need to see what is left of the minuend. So the program branches on the first hyperplane (index 4) of the first subtrahend, and it looks at the side of the hyperplane on which the first subtrahend is not (index 8). The first halfplane happens to be `x < 0.5`, so now the program looks at `x > 0.5` to see whether this part of the minuend is non-empty. It is, so the level is increased.

Second pass through while loop:
```
level == 1
counter == [1, 0]
INDICES = [0, 1, 2, 3, 8]
```
The program is at first pass at level 1. The two hyperplanes (indices 6 and 7) of the remaining subtrahend (the upper right box) is intersected with the right part of the minuend (right of 0.5). The intersection is non-empty, so there is a need to see what is left of the right part of the minuend. The program branches on the first hyperplane (index 6) of the remaining subtrahend, and it looks on the side on which the subtrahend is not (index 10). This halfplane happens to be `x > 0.5`, so now the program looks at `x < 0.5` to see whether this part of the right part of the minuend is non-empty. It is not, so the level remains at the current value.

Third pass through while loop:
```
level == 1
counter == [1, 1]
INDICES == [0, 1, 2, 3, 8, 10]
```
The program is at second pass at level 1. Now it needs to look at the other side of the first hyperplane of the second subtrahend (`x > 0.5`, index 6), and on the side of the second hyperplane (index 7) on which the second subtrahend is not (`y < 0.5`, index 11), to see if this part of the minuend is non-empty. It is non-empty, so the program can continue. But since the program now is at the highest possible level (there are no more subtrahends to subtract) the current set of halfplanes (indices [1, 2, 3, 8, 6, 11], representing the polytope `0.5 < x < 1.0` and ´0.0 < y < 0.5`) is a subset of the resulting region difference. Therefore, it is added to the result.

Fourth pass through while loop:
```
level == 1
counter == [1, 2]
INDICES == [0, 1, 2, 3, 8, 6, 11]
```
The program is done at this level, so it pops out of this level by decrementing the level and resetting the counter and the indices to a previous state, and then advances the state to the next. The next state should check `x < 0.5` and `y > 0.5` (indices 4 and 9). This part of the minuend is non-empty, so the program increments the level to see whether there are any subtrahends left to subtract.

Fifth pass through the while loop:
```
level == 1
counter == [2, 0]
INDICES == [0, 1, 2, 3, 4, 9]
```
The program is at the first pass at level 1, but the current set of halfplanes does not intersect with the remaining subtrahend. Therefore the program enters the `if` statement on removed line 2078. Here things start to go wrong. The conclusion at this point, that the current set of indices represents a subset of the resulting difference, is correct. But then the program should back out of the current level (nothing more here to do) and advance the state. However, what happens is that the level is decremented and the program will next check `0.0 < x < 0.5` and `0.0 < y < 0.5` (indices 0, 1, 2, 3, 4, 5, and 10). This is a non-empty set, and it does not intersect with the second subtrahend, so the first subtrhend is added to the result. This is of course not good.

The problem is that the program does not pop out of the current level in a correct way. The correct way to pop a level is implemented in the `else` statement on line 2112. To get there, `INDICES` needs to represent an empty set, and `counter[level]` must indicate that the program is done at the current level (`counter[level]` must be at least `mi[level]`).